src/rpc/RPCHelpers.cpp

```diff
@@ -426,49 +503,154 @@ ledgerInfoToBlob(ripple::LedgerInfo const& info, bool includeHash)
+    auto parsedCursor =
+        parseAccountCursor(backend, sequence, jsonCursor, accountID, yield);
+
+    if (!parsedCursor)
+        return Status(ripple::rpcINVALID_PARAMS, "Malformed cursor");
+
+    auto cursor = AccountCursor({beast::zero, 0});
+
+    auto [hexCursor, startHint] = *parsedCursor;
+
     auto const rootIndex = ripple::keylet::ownerDir(accountID);
     auto currentIndex = rootIndex;
 
     std::vector<ripple::uint256> keys;
-    std::optional<ripple::uint256> nextCursor = {};
+    keys.reserve(limit);
 
     auto start = std::chrono::system_clock::now();
-    for (;;)
+
+    // If startAfter is not zero try jumping to that page using the hint
+    if (hexCursor.isNonZero())
     {
-        auto ownedNode =
-            backend.fetchLedgerObject(currentIndex.key, sequence, yield);
+        auto const hintIndex = ripple::keylet::page(rootIndex, startHint);
+        auto hintDir =
+            backend.fetchLedgerObject(hintIndex.key, sequence, yield);
 
-        if (!ownedNode)
+        if (hintDir)
         {
-            break;
+            ripple::SerialIter it{hintDir->data(), hintDir->size()};
+            ripple::SLE sle{it, hintIndex.key};
+
+            for (auto const& key : sle.getFieldV256(ripple::sfIndexes))
+            {
+                if (key == hexCursor)
+                {
+                    // We found the hint, we can start here
+                    currentIndex = hintIndex;
+                    break;
+                }
+            }
         }
 
-        ripple::SerialIter it{ownedNode->data(), ownedNode->size()};
-        ripple::SLE dir{it, currentIndex.key};
+        bool found = false;
+        for (;;)
+        {
+            auto const ownerDir =
+                backend.fetchLedgerObject(hintIndex.key, sequence, yield);
+
+            if (!ownerDir)
+                return Status(
+                    ripple::rpcINVALID_PARAMS, "Owner directory not found");
 
-        for (auto const& key : dir.getFieldV256(ripple::sfIndexes))
+            ripple::SerialIter it{ownerDir->data(), ownerDir->size()};
+            ripple::SLE sle{it, hintIndex.key};
+
+            for (auto const& key : sle.getFieldV256(ripple::sfIndexes))
+            {
+                if (!found)
+                {
+                    if (key == hexCursor)
+                        found = true;
+                }
+                else
+                {
+                    keys.push_back(key);
+
+                    if (--limit == 0)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            auto const uNodeNext = sle.getFieldU64(ripple::sfIndexNext);
+
+            if (limit == 0 && uNodeNext != 0)
+            {
+                cursor = AccountCursor({keys.back(), uNodeNext});
+                break;
+            }
+
+            if (uNodeNext == 0)
+                break;
+
+            currentIndex = ripple::keylet::page(rootIndex, uNodeNext);
+        }
```

Comment:
This loop is almost exactly the same as the loop in the below `else` body. Can we merge them?